zone_affinity.go

```diff
@@ -0,0 +1,169 @@
+package cluster
+
+import (
+	"io/ioutil"
+	"net/http"
+	"os"
+	"time"
+)
+
+const overrideEnvVar = "AVAILABILITY_ZONE"
+
+// ZoneAffinityBalancer makes a best effort to pair up consumers with partitions
+// where the leader is in the same zone.  Doing so can help from a performance
+// perspective by minimizing round trip latency, and it can also help keep costs
+// down when running in AWS by avoid cross-AZ data transfer.  Partitions are
+// spread evenly across consumers.
+//
+// The consumers communicate their zones via UserData.  If the zone is missing
+// or cannot be determined, then the data will be empty.  The zone can also be
+// set manually using the AVAILABILITY_ZONE environment variable...this is
+// useful e.g. when doing local testing.
+type ZoneAffinityBalancer struct {
+	zone []byte
+}
+
+func NewZoneAffinityBalancer() *ZoneAffinityBalancer {
+	return &ZoneAffinityBalancer{zone: findZone()}
+}
+
+func (za *ZoneAffinityBalancer) Balance(members []Member, partitions []Partition) map[string][]int32 {
```

Comment:
Might be cool for a metric here for assignments in zone, or you could even report a ratio of assignments in zone vs. out of zone, that would be neat.